### PR TITLE
[SPARK-11120] Allow sane default number of executor failures when dynamically allocating in YARN

### PR DESCRIPTION
**Pull request by ryan-williams**
I also added some information to container-failure error msgs about what host they failed on, which would have helped me identify the problem that lead me to this JIRA and PR sooner.

### CODE REVIEW

File: yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala [hunk 1]
```diff
@@ -62,10 +62,23 @@ private[spark] class ApplicationMaster(
...
+  // Default to numExecutors * 2 (maxExecutors in the case that we are
+  // dynamically allocating executors), with minimum of 3.
+  private val maxNumExecutorFailures =
+    sparkConf.getInt("spark.yarn.max.executor.failures",
+      sparkConf.getInt("spark.yarn.max.worker.failures",
+        math.max(
+          3,
+          2 * sparkConf.getInt(
+            if (Utils.isDynamicAllocationEnabled(sparkConf))
+              "spark.dynamicAllocation.maxExecutors"
```
Comment (jerryshao) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala]: If "spark.dynamicAllocation.maxExecutors" is not set, it is still 3 as minimum. So from your change, user have to configure "spark.dynamicAllocation.maxExecutors" to increase max failure attempts if dynamic allocation is enabled. So in that way why not directly configure "spark.yarn.max.worker.failures", that will be more straightforward?

Comment (ryan-williams) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala]: To be clear, this change does not place any additional requirements on a user to set `maxExecutors` to get sane dynamic allocation (DA) default behavior.

It merely alleviates one class of "gotcha" that caused me some trouble this week: when setting standard DA params, the `val maxNumExecutorFailures` here becomes `3` by default, which does not seem sensible for apps that are going up to many 100s of executors.

It seems to me that the extant `math.max(sparkConf.getInt("spark.executor.instances", 0) *  2, 3)` expression is not _intentionally_ making DA apps have a limit of `3` failures, but that it simply wasn't taking into account the fact that `spark.executor.instances` is not set in DA mode.

It's true that we could also "resolve" this by declaring `spark.yarn.max.worker.failures` to be yet another configuration param that must be set to a non-default value in order to get sane DA behavior.

Off the top of my head, there is already one param (`spark.shuffle.service.enabled=true`) that is not named in a way that suggests that it is important for DA apps to set, and we could make `spark.yarn.max.worker.failures` a second.

My belief is that it would be better to not require yet another parameter (especially one that is not named in a way that makes it obvious that it is or could be important for DA to not fail in unexpected ways) for sane DA behavior, but to just fix the clearly-inadvertently-missed setting of a good default value here.

File: yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala [hunk 2]
```diff
@@ -62,10 +62,24 @@ private[spark] class ApplicationMaster(
...
 
-  // Default to numExecutors * 2, with minimum of 3
-  private val maxNumExecutorFailures = sparkConf.getInt("spark.yarn.max.executor.failures",
-    sparkConf.getInt("spark.yarn.max.worker.failures",
-      math.max(sparkConf.getInt("spark.executor.instances", 0) *  2, 3)))
+  // Default to numExecutors * 2 (maxExecutors in the case that we are
+  // dynamically allocating executors), with minimum of 3.
+  private val maxNumExecutorFailures =
+    sparkConf.getInt("spark.yarn.max.executor.failures",
+      sparkConf.getInt("spark.yarn.max.worker.failures",
```
Comment (vanzin) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala]: This is not your fault, but since you're touching it... this should be added to `SparkConf.configsWithAlternatives` instead.

File: yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala [hunk 3]
```diff
@@ -62,10 +62,24 @@ private[spark] class ApplicationMaster(
...
+        math.max(
+          3,
+          2 * sparkConf.getInt(
```
Comment (vanzin) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala]: I think it would be easier to read if this were pulled into another variable.

    private val maxNumExecutorFailures = {
      val executorCount = sparkConf.getInt(...)
      val defaultFailures = math.max(3, executorCount)
      sparkConf.getInt(...)
    }

**Comment by vanzin** [comment]
Change seems fine to me, I just think the code could be cleaned up a bit to help readability.

File: yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala (hunk 2 above)
Comment (ryan-williams) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala]: ah, hadn't seen that, done

File: yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala (hunk 3 above)
Comment (ryan-williams) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala]: cool, done with some slightly more verbose `val`-names, lmk if you want me to change

File: yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala [hunk 4]
```diff
@@ -62,10 +62,26 @@ private[spark] class ApplicationMaster(
...
+          "spark.dynamicAllocation.maxExecutors"
+        } else {
+          "spark.executor.instances"
+        },
+        0
+      )
+
+    val defaultMaxNumExecutorFailures = math.max(3, 2 * effectiveNumExecutors)
+
+    sparkConf.getInt(
```
Comment (vanzin) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala]: minor nit, but this fits in one line now.

**Comment by vanzin** [comment]
LGTM pending tests.

File: yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala (hunk 4 above)
Comment (ryan-williams) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala]: OK I addressed this so the tests will run one more time

**Comment by ryan-williams** [comment]
Test failure above looks like a jenkins⟷git issue?

**Comment by vanzin** [comment]
yeah. don't worry about it, no need to run the tests again.

File: yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala [hunk 5]
```diff
@@ -62,10 +62,23 @@ private[spark] class ApplicationMaster(
     .asInstanceOf[YarnConfiguration]
   private val isClusterMode = args.userClass != null
 
-  // Default to numExecutors * 2, with minimum of 3
-  private val maxNumExecutorFailures = sparkConf.getInt("spark.yarn.max.executor.failures",
-    sparkConf.getInt("spark.yarn.max.worker.failures",
-      math.max(sparkConf.getInt("spark.executor.instances", 0) *  2, 3)))
+  // Default to numExecutors * 2 (maxExecutors in the case that we are
+  // dynamically allocating executors), with minimum of 3.
```
Comment (andrewor14) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala]: this is hard to read. Can you rephrase:
```
// If dynamic allocation is enabled, default to the upper limit.
// Otherwise, default to 2 * numExecutors with a minimum of 3.
```

File: yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala [hunk 6]
```diff
@@ -62,10 +62,23 @@ private[spark] class ApplicationMaster(
...
+        if (Utils.isDynamicAllocationEnabled(sparkConf)) {
+          "spark.dynamicAllocation.maxExecutors"
+        } else {
+          "spark.executor.instances"
+        },
+        0
+      )
+
+    val defaultMaxNumExecutorFailures = math.max(3, 2 * effectiveNumExecutors)
+
```
Comment (andrewor14) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala]: can you rewrite this so it's more readable:
```
val defaultKey =
  if (Utils.isDynamicAllocationEnabled(sparkConf)) {
    "spark.dynamicAllocation.maxExecutors"
  } else {
    "spark.executor.instances"
  }
val effectiveNumExecutors = sparkConf.getInt(defaultKey, 0)
val defaultMaxNumExeuctorFailures = math.max(3, 2 * effectiveNumExecutors)
sparkConf.getInt(..., defaultMaxNumExecutorFailures)
```

File: yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala (hunk 5 above)
Comment (ryan-williams) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala]: done (a version of that anyway)

File: yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala (hunk 6 above)
Comment (ryan-williams) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala]: done

File: yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala [hunk 7]
```diff
@@ -62,10 +62,19 @@ private[spark] class ApplicationMaster(
...
+  // Default to twice the number of executors (twice the maximum number of executors if dynamic
+  // allocation is enabled), with a minimum of 3.
+  val defaultKey =
+    if (Utils.isDynamicAllocationEnabled(sparkConf)) {
+      "spark.dynamicAllocation.maxExecutors"
+    } else {
+      "spark.executor.instances"
+    }
+  val effectiveNumExecutors = sparkConf.getInt(defaultKey, 0)
+  val defaultMaxNumExecutorFailures = math.max(3, 2 * effectiveNumExecutors)
```
Comment (andrewor14) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala]: you addressed my comment quickly, but these should not be public or outside the scope of `maxNumExecutorFailures`. What I meant is more like:
```
// Default to twice ...
private val maxNumExecutorFailures: Int = {
  val defaultKey = ...
  sparkConf.getInt(...)
}
```

**Comment by andrewor14** [comment]
Looks good. Can you add a short test?

File: yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala (hunk 7 above)
Comment (ryan-williams) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala]: woops, didn't notice that you still wanted that scoping, fixed

**Comment by ryan-williams** [comment]
not super excited to write a test for this, tbh. what kind of test are you imagining? `ApplicationMaster*Suite` doesn't even exist atm...

**Comment by andrewor14** [comment]
I was thinking just checking the value of `maxNumExecutorFailures` with and without dynamic allocation being set and asserting that it's as we expect. Though this change is pretty straightforward so we can add the test separately if you prefer.

retest this please

File: yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala [hunk 8]
```diff
@@ -430,17 +430,20 @@ private[yarn] class YarnAllocator(
...
       val containerId = completedContainer.getContainerId
       val alreadyReleased = releasedContainers.remove(containerId)
+      val hostOpt = allocatedContainerToHostMap.get(containerId)
+      val onHostStr = hostOpt.map(host => s" on host: $host").getOrElse("")
       val exitReason = if (!alreadyReleased) {
         // Decrement the number of executors running. The next iteration of
         // the ApplicationMaster's reporting thread will take care of allocating.
         numExecutorsRunning -= 1
-        logInfo("Completed container %s (state: %s, exit status: %s)".format(
+        logInfo("Completed container %s%s (state: %s, exit status: %s)".format(
```
Comment (srowen) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala]: You're welcome to convert this to string interpolation now. (Just if you find you need to make other changes; not worth it only for this.)

Comment (ryan-williams) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala]: Yea I originally did some more aggressive converting to interpolated-strings here because that seems like a better way to do things in general, but such strings often make for absurdly long lines, and I don't have a solution I like to that (e.g. breaking onto multiple lines with `+`s seems gross), so I think I reverted that here and went for the more minimal change.

**Comment by ryan-williams** [comment]
Seems like this did not attempt to retest, @andrewor14?

File: yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala (hunk 8 above)
Comment (srowen) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala]: (FWIW I don't mind `s"..." + s"..."` myself.) I'll trigger a test.

**Comment by ryan-williams** [comment]
not sure how [this test failure](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1929/testReport/org.apache.spark.launcher/LauncherBackendSuite/standalone_client__launcher_handle/) could be related to this change

**Comment by andrewor14** [comment]
retest this please

**Comment by andrewor14** [comment]
Actually, I'm just going to merge this since all the YARN / dynamic allocation tests already passed. The one that failed is a known flaky test. This is going into master. Thanks @ryan-williams.
